// symbol-sdk/dist/src/model/mosaic/MosaicFlags.js
"use strict";
/*
 * Copyright 2018 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicFlags = void 0;
const catbuffer_typescript_1 = require("catbuffer-typescript");
/**
 * Mosaic flags model
 */
class MosaicFlags {
    /**
     * @param flags
     */
    constructor(flags) {
        this.supplyMutable = (flags & catbuffer_typescript_1.MosaicFlagsDto.SUPPLY_MUTABLE) !== 0;
        this.transferable = (flags & catbuffer_typescript_1.MosaicFlagsDto.TRANSFERABLE) !== 0;
        this.restrictable = (flags & catbuffer_typescript_1.MosaicFlagsDto.RESTRICTABLE) !== 0;
        this.revokable = (flags & catbuffer_typescript_1.MosaicFlagsDto.REVOKABLE) !== 0;
    }
    /**
     * Static constructor function with default parameters
     * @returns {MosaicFlags}
     * @param supplyMutable
     * @param transferable
     * @param restrictable
     * @param revokable
     */
    static create(supplyMutable, transferable, restrictable = false, revokable = false) {
        const flags = this.toFlag({
            supplyMutable: supplyMutable,
            transferable: transferable,
            restrictable: restrictable,
            revokable: revokable,
        });
        return new MosaicFlags(flags);
    }
    /**
     * Get mosaic flag value in number
     * @returns {number}
     */
    getValue() {
        return MosaicFlags.toFlag(this);
    }
    /**
     * Create DTO object
     */
    toDTO() {
        return {
            flags: this.getValue(),
        };
    }
    /**
     * It "adds up" individual flags into a bit wise number flag.
     *
     * @param supplyMutable - if the supply is mutable. First flag.
     * @param transferable - if the balance can be transferred. Second flag.
     * @param restrictable - if the transaction can be restricted. Third flag.
     * @param revokable - if the balance can be revoked. Fourth flag.
     * @private
     */
    static toFlag({ supplyMutable, transferable, restrictable, revokable, }) {
        return (supplyMutable ? 1 : 0) + (transferable ? 2 : 0) + (restrictable ? 4 : 0) + (revokable ? 8 : 0);
    }
}
exports.MosaicFlags = MosaicFlags;
