// symbol-sdk/dist/src/model/transaction/MosaicSupplyRevocationTransaction.js
"use strict";
/*
 * Copyright 2021 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicSupplyRevocationTransaction = void 0;
const catbuffer_typescript_1 = require("catbuffer-typescript");
const core_1 = require("../../core");
const format_1 = require("../../core/format");
const UnresolvedMapping_1 = require("../../core/utils/UnresolvedMapping");
const PublicAccount_1 = require("../account/PublicAccount");
const mosaic_1 = require("../mosaic");
const UInt64_1 = require("../UInt64");
const Deadline_1 = require("./Deadline");
const Transaction_1 = require("./Transaction");
const TransactionType_1 = require("./TransactionType");
const TransactionVersion_1 = require("./TransactionVersion");
/**
 * Creators of a revokable mosaic will be able to recall any and all balances from any holders. Holders of these mosaics implicitly place trust in the issuer.
 * The mosaic issuer can revoke and recall balances using this transaction.
 */
class MosaicSupplyRevocationTransaction extends Transaction_1.Transaction {
    /**
     * @param networkType
     * @param version
     * @param deadline
     * @param maxFee
     * @param sourceAddress
     * @param mosaic
     * @param signature
     * @param signer
     * @param transactionInfo
     */
    constructor(networkType, version, deadline, maxFee, 
    /**
     * Address from which tokens should be revoked.
     */
    sourceAddress, 
    /**
     * Revoked mosaic and amount.
     */
    mosaic, signature, signer, transactionInfo) {
        super(TransactionType_1.TransactionType.MOSAIC_SUPPLY_REVOCATION, networkType, version, deadline, maxFee, signature, signer, transactionInfo);
        this.sourceAddress = sourceAddress;
        this.mosaic = mosaic;
    }
    /**
     * Create a mosaic supply revocation transaction object
     * @param deadline - The deadline to include the transaction.
     * @param sourceAddress - Address from which tokens should be revoked.
     * @param mosaic - Revoked mosaic and amount.
     * @param networkType - The network type.
     * @param maxFee - (Optional) Max fee defined by the sender
     * @param signature - (Optional) Transaction signature
     * @param signer - (Optional) Signer public account
     * @returns {MosaicSupplyRevocationTransaction}
     */
    static create(deadline, sourceAddress, mosaic, networkType, maxFee = new UInt64_1.UInt64([0, 0]), signature, signer) {
        return new MosaicSupplyRevocationTransaction(networkType, TransactionVersion_1.TransactionVersion.MOSAIC_SUPPLY_REVOCATION, deadline, maxFee, sourceAddress, mosaic, signature, signer);
    }
    /**
     * Create a transaction object from payload
     * @param {string} payload Binary payload
     * @param {Boolean} isEmbedded Is embedded transaction (Default: false)
     * @returns {Transaction | InnerTransaction}
     */
    static createFromPayload(payload, isEmbedded = false) {
        const builder = isEmbedded
            ? catbuffer_typescript_1.EmbeddedMosaicSupplyRevocationTransactionBuilder.loadFromBinary(format_1.Convert.hexToUint8(payload))
            : catbuffer_typescript_1.MosaicSupplyRevocationTransactionBuilder.loadFromBinary(format_1.Convert.hexToUint8(payload));
        const mosaicBuilder = builder.getMosaic();
        const id = new UInt64_1.UInt64(mosaicBuilder.mosaicId.unresolvedMosaicId).toHex();
        const mosaic = new mosaic_1.Mosaic(UnresolvedMapping_1.UnresolvedMapping.toUnresolvedMosaic(id), new UInt64_1.UInt64(mosaicBuilder.amount.amount));
        const signerPublicKey = format_1.Convert.uint8ToHex(builder.getSignerPublicKey().publicKey);
        const networkType = builder.getNetwork().valueOf();
        const signature = Transaction_1.Transaction.getSignatureFromPayload(payload, isEmbedded);
        const deadline = isEmbedded
            ? Deadline_1.Deadline.createEmtpy()
            : Deadline_1.Deadline.createFromDTO(builder.getDeadline().timestamp);
        const transaction = MosaicSupplyRevocationTransaction.create(deadline, UnresolvedMapping_1.UnresolvedMapping.toUnresolvedAddress(format_1.Convert.uint8ToHex(builder.getSourceAddress().unresolvedAddress)), mosaic, networkType, isEmbedded ? new UInt64_1.UInt64([0, 0]) : new UInt64_1.UInt64(builder.fee.amount), signature, signerPublicKey.match(`^[0]+$`) ? undefined : PublicAccount_1.PublicAccount.createFromPublicKey(signerPublicKey, networkType));
        return isEmbedded ? transaction.toAggregate(PublicAccount_1.PublicAccount.createFromPublicKey(signerPublicKey, networkType)) : transaction;
    }
    /**
     * @internal
     * @returns {TransactionBuilder}
     */
    createBuilder() {
        const mosaicBuilder = new catbuffer_typescript_1.UnresolvedMosaicBuilder(new catbuffer_typescript_1.UnresolvedMosaicIdDto(this.mosaic.id.id.toDTO()), new catbuffer_typescript_1.AmountDto(this.mosaic.amount.toDTO()));
        return new catbuffer_typescript_1.MosaicSupplyRevocationTransactionBuilder(this.getSignatureAsBuilder(), this.getSignerAsBuilder(), this.versionToDTO(), this.networkType.valueOf(), this.type.valueOf(), new catbuffer_typescript_1.AmountDto(this.maxFee.toDTO()), new catbuffer_typescript_1.TimestampDto(this.deadline.toDTO()), new catbuffer_typescript_1.UnresolvedAddressDto(this.sourceAddress.encodeUnresolvedAddress(this.networkType)), mosaicBuilder);
    }
    /**
     * @internal
     * @returns {EmbeddedTransactionBuilder}
     */
    toEmbeddedTransaction() {
        const mosaicBuilder = new catbuffer_typescript_1.UnresolvedMosaicBuilder(new catbuffer_typescript_1.UnresolvedMosaicIdDto(this.mosaic.id.id.toDTO()), new catbuffer_typescript_1.AmountDto(this.mosaic.amount.toDTO()));
        return new catbuffer_typescript_1.EmbeddedMosaicSupplyRevocationTransactionBuilder(this.getSignerAsBuilder(), this.versionToDTO(), this.networkType.valueOf(), this.type.valueOf(), new catbuffer_typescript_1.UnresolvedAddressDto(this.sourceAddress.encodeUnresolvedAddress(this.networkType)), mosaicBuilder);
    }
    /**
     * @internal
     * @param statement Block receipt statement
     * @param aggregateTransactionIndex Transaction index for aggregated transaction
     * @returns {MosaicSupplyRevocationTransaction}
     */
    resolveAliases(statement, aggregateTransactionIndex = 0) {
        const transactionInfo = this.checkTransactionHeightAndIndex();
        return core_1.DtoMapping.assign(this, {
            sourceAddress: statement.resolveAddress(this.sourceAddress, transactionInfo.height.toString(), transactionInfo.index, aggregateTransactionIndex),
            mosaic: statement.resolveMosaic(this.mosaic, transactionInfo.height.toString(), transactionInfo.index, aggregateTransactionIndex),
        });
    }
    /**
     * @internal
     * Check a given address should be notified in websocket channels
     * @param address address to be notified
     * @returns {boolean}
     */
    shouldNotifyAccount(address) {
        return super.isSigned(address);
    }
}
exports.MosaicSupplyRevocationTransaction = MosaicSupplyRevocationTransaction;
