// symbol-openapi-typescript-fetch-client/src/models/DeploymentDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest
 *
 * The version of the OpenAPI document: 1.0.3
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface DeploymentDTO
 */
export interface DeploymentDTO {
    /**
     * The tool used to create, maintain and deploy the node. Examples: symbol-bootstrap, manual.
     * @type {string}
     * @memberof DeploymentDTO
     */
    deploymentTool: string;
    /**
     * The version of the tool used to create, maintain and deploy the node.
     * @type {string}
     * @memberof DeploymentDTO
     */
    deploymentToolVersion: string;
    /**
     * When was the node last upgraded.
     * @type {string}
     * @memberof DeploymentDTO
     */
    lastUpdatedDate: string;
}

export function DeploymentDTOFromJSON(json: any): DeploymentDTO {
    return DeploymentDTOFromJSONTyped(json, false);
}

export function DeploymentDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): DeploymentDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'deploymentTool': json['deploymentTool'],
        'deploymentToolVersion': json['deploymentToolVersion'],
        'lastUpdatedDate': json['lastUpdatedDate'],
    };
}

export function DeploymentDTOToJSON(value?: DeploymentDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'deploymentTool': value.deploymentTool,
        'deploymentToolVersion': value.deploymentToolVersion,
        'lastUpdatedDate': value.lastUpdatedDate,
    };
}
