// catbuffer-typescript/dist/TransactionTypeDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionTypeDto = void 0;
var TransactionTypeDto;
(function (TransactionTypeDto) {
    TransactionTypeDto[TransactionTypeDto["ACCOUNT_KEY_LINK"] = 16716] = "ACCOUNT_KEY_LINK";
    TransactionTypeDto[TransactionTypeDto["NODE_KEY_LINK"] = 16972] = "NODE_KEY_LINK";
    TransactionTypeDto[TransactionTypeDto["AGGREGATE_COMPLETE"] = 16705] = "AGGREGATE_COMPLETE";
    TransactionTypeDto[TransactionTypeDto["AGGREGATE_BONDED"] = 16961] = "AGGREGATE_BONDED";
    TransactionTypeDto[TransactionTypeDto["VOTING_KEY_LINK"] = 16707] = "VOTING_KEY_LINK";
    TransactionTypeDto[TransactionTypeDto["VRF_KEY_LINK"] = 16963] = "VRF_KEY_LINK";
    TransactionTypeDto[TransactionTypeDto["HASH_LOCK"] = 16712] = "HASH_LOCK";
    TransactionTypeDto[TransactionTypeDto["SECRET_LOCK"] = 16722] = "SECRET_LOCK";
    TransactionTypeDto[TransactionTypeDto["SECRET_PROOF"] = 16978] = "SECRET_PROOF";
    TransactionTypeDto[TransactionTypeDto["ACCOUNT_METADATA"] = 16708] = "ACCOUNT_METADATA";
    TransactionTypeDto[TransactionTypeDto["MOSAIC_METADATA"] = 16964] = "MOSAIC_METADATA";
    TransactionTypeDto[TransactionTypeDto["NAMESPACE_METADATA"] = 17220] = "NAMESPACE_METADATA";
    TransactionTypeDto[TransactionTypeDto["MOSAIC_DEFINITION"] = 16717] = "MOSAIC_DEFINITION";
    TransactionTypeDto[TransactionTypeDto["MOSAIC_SUPPLY_CHANGE"] = 16973] = "MOSAIC_SUPPLY_CHANGE";
    TransactionTypeDto[TransactionTypeDto["MOSAIC_SUPPLY_REVOCATION"] = 17229] = "MOSAIC_SUPPLY_REVOCATION";
    TransactionTypeDto[TransactionTypeDto["MULTISIG_ACCOUNT_MODIFICATION"] = 16725] = "MULTISIG_ACCOUNT_MODIFICATION";
    TransactionTypeDto[TransactionTypeDto["ADDRESS_ALIAS"] = 16974] = "ADDRESS_ALIAS";
    TransactionTypeDto[TransactionTypeDto["MOSAIC_ALIAS"] = 17230] = "MOSAIC_ALIAS";
    TransactionTypeDto[TransactionTypeDto["NAMESPACE_REGISTRATION"] = 16718] = "NAMESPACE_REGISTRATION";
    TransactionTypeDto[TransactionTypeDto["ACCOUNT_ADDRESS_RESTRICTION"] = 16720] = "ACCOUNT_ADDRESS_RESTRICTION";
    TransactionTypeDto[TransactionTypeDto["ACCOUNT_MOSAIC_RESTRICTION"] = 16976] = "ACCOUNT_MOSAIC_RESTRICTION";
    TransactionTypeDto[TransactionTypeDto["ACCOUNT_OPERATION_RESTRICTION"] = 17232] = "ACCOUNT_OPERATION_RESTRICTION";
    TransactionTypeDto[TransactionTypeDto["MOSAIC_ADDRESS_RESTRICTION"] = 16977] = "MOSAIC_ADDRESS_RESTRICTION";
    TransactionTypeDto[TransactionTypeDto["MOSAIC_GLOBAL_RESTRICTION"] = 16721] = "MOSAIC_GLOBAL_RESTRICTION";
    TransactionTypeDto[TransactionTypeDto["TRANSFER"] = 16724] = "TRANSFER";
})(TransactionTypeDto = exports.TransactionTypeDto || (exports.TransactionTypeDto = {}));
